### PR TITLE
Feature/identity update

### PR DESCRIPTION
**Pull request by mikron**

### CODE REVIEW

File: identity-api/identity-api-common/src/main/java/com/sflpro/identity/api/common/dtos/principal/PrincipalNameBusyExceptionDto.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+import com.sflpro.identity.api.common.dtos.IdentityApiExceptionDto;
+import com.sflpro.identity.api.common.dtos.NonLoggableException;
+
+/**
+ * Company: SFL LLC
+ * Created on 17/02/2018
+ *
+ * @author Davit Harutyunyan
+ */
+public class PrincipalNameBusyExceptionDto extends IdentityApiExceptionDto implements NonLoggableException {
```
Comment (sfl-sonar) [inline identity-api/identity-api-common/src/main/java/com/sflpro/identity/api/common/dtos/principal/PrincipalNameBusyExceptionDto.java]: MAJOR This class has 6 parents which is greater than 5 authorized. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AMaximumInheritanceDepth)

File: identity-api/identity-api-common/src/main/java/com/sflpro/identity/api/common/dtos/IdentityApiExceptionDto.java [hunk 2]
```diff
@@ -32,9 +41,9 @@ public IdentityApiExceptionDto(IdentityApiError error, String message, Exception
...
 
-    public IdentityApiExceptionDto(IdentityApiError error,
-                                   String message,
-                                   int responseStatusCode) {
+    public IdentityApiExceptionDto(@JsonProperty("status") IdentityApiError error,
+                                  @JsonProperty("message") String message,
+                                  @JsonProperty("httpStatus") int responseStatusCode) {
         super(message, error.getResponseHttpStatus());
         this.applicationErrorCode = error.getErrorCode();
         this.message = message == null ? error.getDefaultMessage() == null ? error.toString() : error.getDefaultMessage() : message;
```
Comment (sfl-sonar) [inline identity-api/identity-api-common/src/main/java/com/sflpro/identity/api/common/dtos/IdentityApiExceptionDto.java]: MAJOR Extract this nested ternary operation into an independent statement. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3358)

File: identity-core/identity-core-services-impl/src/main/java/com/sflpro/identity/core/services/identity/IdentityServiceImpl.java [hunk 3]
```diff
@@ -87,6 +93,61 @@ public Identity create(final IdentityCreationRequest identityCreationRequest) {
...
+        Assert.notNull(updateRequest, "updateRequest cannot be null");
+        logger.debug("Updating identity by id {}", identityId);
+        Identity identity = get(identityId);
+
+        identity.setDescription(updateRequest.getDescription());
+        // Change secret
+        if (secretHashHelper.isSecretCorrect(updateRequest.getSecret(), identity.getSecret())) {
+            changeSecret(identity, updateRequest.getNewSecret());
+        } else {
+            throw new AuthenticationServiceException("Invalid credentials"); // TODO check with Mr. Smith
```
Comment (sfl-sonar) [inline identity-core/identity-core-services-impl/src/main/java/com/sflpro/identity/core/services/identity/IdentityServiceImpl.java]: INFO Complete the task associated to this TODO comment. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1135)

File: identity-core/identity-core-services-impl/src/main/java/com/sflpro/identity/core/services/identity/IdentityServiceImpl.java [hunk 4]
```diff
@@ -87,6 +93,61 @@ public Identity create(final IdentityCreationRequest identityCreationRequest) {
...
+        }
+        // Update principals
+        // TODO every time we are updating MAIN statues to SECONDARY, which is weird
```
Comment (sfl-sonar) [inline identity-core/identity-core-services-impl/src/main/java/com/sflpro/identity/core/services/identity/IdentityServiceImpl.java]: INFO Complete the task associated to this TODO comment. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1135)

File: identity-core/identity-core-services-impl/src/main/java/com/sflpro/identity/core/services/identity/IdentityServiceImpl.java [hunk 5]
```diff
@@ -68,7 +70,11 @@ public Identity get(final String identityId) {
         return identity;
     }
 
+    /**
+     * {@inheritDoc}
+     */
     @Override
+    @Deprecated
     public Identity create(final IdentityCreationRequest identityCreationRequest) {
```
Comment (sfl-sonar) [inline identity-core/identity-core-services-impl/src/main/java/com/sflpro/identity/core/services/identity/IdentityServiceImpl.java]: MAJOR Add the missing @deprecated Javadoc tag. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AMissingDeprecatedCheck)
INFO Do not forget to remove this deprecated code someday. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1133)

File: identity-core/identity-core-services-impl/src/main/java/com/sflpro/identity/core/services/identity/IdentityServiceImpl.java [hunk 6]
```diff
@@ -87,6 +93,61 @@ public Identity create(final IdentityCreationRequest identityCreationRequest) {
...
+
+        logger.trace("Complete updating identity by id {}.", identityId);
+        return identity;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    /*@Override
+    public Identity checkMailAvailability(final IdentityCheckPrincipalRequest request) {
```
Comment (sfl-sonar) [inline identity-core/identity-core-services-impl/src/main/java/com/sflpro/identity/core/services/identity/IdentityServiceImpl.java]: MAJOR This block of commented-out lines of code should be removed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

File: identity-api/identity-api-facade/src/main/java/com/sflpro/identity/api/endpoints/IdentityEndpoint.java [hunk 7]
```diff
@@ -60,6 +67,7 @@ public IdentityDto get(@NotNull @PathParam("identityId") final String identityId
     @ApiOperation("Creates identity")
     @PUT
     @Transactional
+    @Deprecated
     public IdentityCreationResponseDto create(@NotNull @Valid final IdentityCreationRequestDto identityCreationRequestDto) {
```
Comment (sfl-sonar) [inline identity-api/identity-api-facade/src/main/java/com/sflpro/identity/api/endpoints/IdentityEndpoint.java]: MAJOR Add the missing @deprecated Javadoc tag. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AMissingDeprecatedCheck)
INFO Do not forget to remove this deprecated code someday. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1133)

File: identity-api/identity-api-facade/src/main/java/com/sflpro/identity/api/endpoints/IdentityEndpoint.java [hunk 8]
```diff
@@ -74,18 +82,46 @@ public IdentityCreationResponseDto create(@NotNull @Valid final IdentityCreation
...
-        return new IdentityUpdateResponseDto();
+    public IdentityDto delete(@NotNull @PathParam("identityId") final String identityId) {
+        return new IdentityDto();
     }
 
+    /*@ApiOperation("Check principal's name availability by type and name")
+    @POST
+    @Path(("/check-principal"))
+    @Transactional(readOnly = true)
+    public IdentityDto checkPrincipalAvailability(@NotNull @Valid final IdentityCheckPrincipalRequestDto requestDto) {
```
Comment (sfl-sonar) [inline identity-api/identity-api-facade/src/main/java/com/sflpro/identity/api/endpoints/IdentityEndpoint.java]: MAJOR This block of commented-out lines of code should be removed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

File: identity-api/identity-api-common/src/main/java/com/sflpro/identity/api/common/dtos/auth/AuthenticationExceptionDto.java [hunk 9]
```diff
@@ -10,17 +11,17 @@
  *
  * @author Davit Harutyunyan
  */
-public class AuthenticationExceptionDtoDto extends IdentityApiExceptionDto implements NonLoggableException {
+public class AuthenticationExceptionDto extends IdentityApiExceptionDto implements NonLoggableException {
```
Comment (sfl-sonar) [inline identity-api/identity-api-common/src/main/java/com/sflpro/identity/api/common/dtos/auth/AuthenticationExceptionDto.java]: MAJOR This class has 6 parents which is greater than 5 authorized. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AMaximumInheritanceDepth)

File: identity-core/identity-core-services/src/main/java/com/sflpro/identity/core/services/principal/PrincipalNameBusyException.java [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+ * Company: SFL LLC
+ * Created on 17/02/2018
+ *
+ * @author Davit Harutyunyan
+ */
+public class PrincipalNameBusyException extends IdentityServiceException {
+
+    private PrincipalType principalType;
+
+    private String name;
```
Comment (sfl-sonar) [inline identity-core/identity-core-services/src/main/java/com/sflpro/identity/core/services/principal/PrincipalNameBusyException.java]: MINOR Make this "name" field final. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1165)

File: identity-core/identity-core-services/src/main/java/com/sflpro/identity/core/services/principal/PrincipalNameBusyException.java [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+
+/**
+ * Company: SFL LLC
+ * Created on 17/02/2018
+ *
+ * @author Davit Harutyunyan
+ */
+public class PrincipalNameBusyException extends IdentityServiceException {
+
+    private PrincipalType principalType;
```
Comment (sfl-sonar) [inline identity-core/identity-core-services/src/main/java/com/sflpro/identity/core/services/principal/PrincipalNameBusyException.java]: MAJOR Remove this unused "principalType" private field. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1068)
MINOR Make this "principalType" field final. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1165)

File: identity-core/identity-core-services/src/main/java/com/sflpro/identity/core/services/identity/IdentityService.java [hunk 12]
```diff
@@ -26,20 +29,41 @@
      * @param identityCreationRequest identity creation request model
      * @return id and details of the created identity
      */
+    @Deprecated
+    @Transactional
     Identity create(final IdentityCreationRequest identityCreationRequest);
```
Comment (sfl-sonar) [inline identity-core/identity-core-services/src/main/java/com/sflpro/identity/core/services/identity/IdentityService.java]: MAJOR Add the missing @deprecated Javadoc tag. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AMissingDeprecatedCheck)
INFO Do not forget to remove this deprecated code someday. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1133)

File: identity-core/identity-core-services/src/main/java/com/sflpro/identity/core/services/identity/IdentityService.java [hunk 13]
```diff
@@ -26,20 +29,41 @@
...
+    @Transactional
+    Identity update(final String identityId, final IdentityUpdateRequest updateRequest) throws AuthenticationServiceException;
+
+    /**
+     * Gets principal by type mail name
+     *
+     * @param identityCheckPrincipalRequest check mail request
+     * @return details of the campaign
+     */
+    /*Identity checkMailAvailability(final IdentityCheckPrincipalRequest identityCheckPrincipalRequest);*/
```
Comment (sfl-sonar) [inline identity-core/identity-core-services/src/main/java/com/sflpro/identity/core/services/identity/IdentityService.java]: MAJOR This block of commented-out lines of code should be removed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

File: identity-core/identity-core-services/src/main/java/com/sflpro/identity/core/services/principal/PrincipalService.java [hunk 14]
```diff
@@ -23,19 +20,9 @@
...
-    /**
-     * Gets principal by identity
-     *
+    /** Updates existing principal or inserts a new one if principal absent
      * @param identity identity
-     * @return principal details
+     * @param updateRequest update data
      */
-    List<Principal> getByIdentity(final Identity identity);
+    Principal upsert(final Identity identity, final PrincipalUpdateRequest updateRequest);;
```
Comment (sfl-sonar) [inline identity-core/identity-core-services/src/main/java/com/sflpro/identity/core/services/principal/PrincipalService.java]: MINOR Remove this empty statement. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AEmptyStatementUsageCheck)

**Comment by sfl-sonar** [comment]
SonarQube analysis reported 19 issues
* ![MAJOR][MAJOR] 11 major
* ![MINOR][MINOR] 3 minor
* ![INFO][INFO] 5 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [IdentityApiExceptionDto.java#L57](https://github.com/sflpro/identity/blob/c1f240a32c9ad2a92380fefa332da73b89257f72/identity-api/identity-api-common/src/main/java/com/sflpro/identity/api/common/dtos/IdentityApiExceptionDto.java#L57): Extract this nested ternary operation into an independent statement. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3358)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: identity-api/identity-api-facade/src/main/java/com/sflpro/identity/api/endpoints/IdentityEndpoint.java [hunk 15]
```diff
@@ -74,16 +82,27 @@ public IdentityCreationResponseDto create(@NotNull @Valid final IdentityCreation
...
+        }
     }
 
     @ApiOperation("Delete identity")
     @DELETE
     @Path("/{identityId}")
-    public IdentityUpdateResponseDto delete(@NotNull @PathParam("identityId") final String identityId) {
-        return new IdentityUpdateResponseDto();
+    @Deprecated
+    public IdentityDto delete(@NotNull @PathParam("identityId") final String identityId) {
```
Comment (sfl-sonar) [inline identity-api/identity-api-facade/src/main/java/com/sflpro/identity/api/endpoints/IdentityEndpoint.java]: MAJOR Add the missing @deprecated Javadoc tag. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AMissingDeprecatedCheck)
INFO Do not forget to remove this deprecated code someday. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1133)

**Comment by sfl-sonar** [comment]
SonarQube analysis reported 16 issues
* ![MAJOR][MAJOR] 9 major
* ![MINOR][MINOR] 2 minor
* ![INFO][INFO] 5 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [IdentityApiExceptionDto.java#L57](https://github.com/sflpro/identity/blob/799dfad8ea8907f9f19f6b5bee7b27f0cff28f1f/identity-api/identity-api-common/src/main/java/com/sflpro/identity/api/common/dtos/IdentityApiExceptionDto.java#L57): Extract this nested ternary operation into an independent statement. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3358)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: identity-core/identity-core-services-impl/src/main/java/com/sflpro/identity/core/services/principal/PrincipalServiceImpl.java [hunk 16]
```diff
@@ -25,27 +27,57 @@
     private PrincipalRepository principalRepository;
 
     @Override
-    public Principal get(final String name, final PrincipalType type) throws ResourceNotFoundException {
+    public Principal get(final PrincipalType type, final String name) throws ResourceNotFoundException {
```
Comment (sfl-sonar) [inline identity-core/identity-core-services-impl/src/main/java/com/sflpro/identity/core/services/principal/PrincipalServiceImpl.java]: MINOR Remove the declaration of thrown exception 'com.sflpro.identity.core.services.ResourceNotFoundException' which is a runtime exception. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)

**Comment by sfl-sonar** [comment]
SonarQube analysis reported 17 issues
* ![MAJOR][MAJOR] 9 major
* ![MINOR][MINOR] 3 minor
* ![INFO][INFO] 5 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [IdentityApiExceptionDto.java#L57](https://github.com/sflpro/identity/blob/124d6bd4728f660840f2a5dc5101bc0486c2dc6d/identity-api/identity-api-common/src/main/java/com/sflpro/identity/api/common/dtos/IdentityApiExceptionDto.java#L57): Extract this nested ternary operation into an independent statement. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3358)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: identity-core/identity-core-services/src/main/java/com/sflpro/identity/core/services/principal/PrincipalNameBusyException.java [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+
+/**
+ * Company: SFL LLC
+ * Created on 17/02/2018
+ *
+ * @author Davit Harutyunyan
+ */
+public class PrincipalNameBusyException extends IdentityServiceException {
+
+    private String name;
```
Comment (sfl-sonar) [inline identity-core/identity-core-services/src/main/java/com/sflpro/identity/core/services/principal/PrincipalNameBusyException.java]: MINOR Make this "name" field final. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1165)

**Comment by sfl-sonar** [comment]
SonarQube analysis reported 14 issues
* ![MAJOR][MAJOR] 8 major
* ![MINOR][MINOR] 1 minor
* ![INFO][INFO] 5 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [IdentityApiExceptionDto.java#L57](https://github.com/sflpro/identity/blob/7fc7a712dae1c8bbc0d254c3b2bbcd6fe2bc7205/identity-api/identity-api-common/src/main/java/com/sflpro/identity/api/common/dtos/IdentityApiExceptionDto.java#L57): Extract this nested ternary operation into an independent statement. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3358)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by sfl-sonar** [comment]
SonarQube analysis reported 13 issues
* ![MAJOR][MAJOR] 8 major
* ![INFO][INFO] 5 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [IdentityApiExceptionDto.java#L57](https://github.com/sflpro/identity/blob/d7bdd6651add152926bca24347db7e815e9d4822/identity-api/identity-api-common/src/main/java/com/sflpro/identity/api/common/dtos/IdentityApiExceptionDto.java#L57): Extract this nested ternary operation into an independent statement. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3358)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Comment by sfl-sonar** [comment]
SonarQube analysis reported 11 issues
* ![MAJOR][MAJOR] 6 major
* ![INFO][INFO] 5 info

Watch the comments in this conversation to review them.

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: identity-api/identity-api-facade/src/main/java/com/sflpro/identity/api/endpoints/IdentityEndpoint.java [hunk 18]
```diff
@@ -74,16 +82,27 @@ public IdentityCreationResponseDto create(@NotNull @Valid final IdentityCreation
...
     @Path("/{identityId}")
-    public IdentityUpdateResponseDto update(@NotNull @PathParam("identityId") final String identityId,
-                                            final IdentityUpdateRequestDto updateRequestDto) {
-        return new IdentityUpdateResponseDto();
+    public IdentityDto update(@NotNull @PathParam("identityId") final String identityId,
+                              final IdentityUpdateRequestDto updateRequestDto) {
+        Assert.notNull(updateRequestDto, "updateRequestDto cannot be null");
+        logger.debug("Updating identity id {} with data :{}...", identityId, updateRequestDto);
+        try {
+            IdentityUpdateRequest checkMailRequest = mapper.map(updateRequestDto, IdentityUpdateRequest.class);
```
Comment (dnavre) [inline identity-api/identity-api-facade/src/main/java/com/sflpro/identity/api/endpoints/IdentityEndpoint.java]: checkMailRequest

File: identity-core/identity-core-services-impl/src/main/java/com/sflpro/identity/core/services/identity/IdentityServiceImpl.java (hunk 3 above)
Comment (dnavre) [reply identity-core/identity-core-services-impl/src/main/java/com/sflpro/identity/core/services/identity/IdentityServiceImpl.java]: Yup, nicely done. checked with me.

File: identity-core/identity-core-services-impl/src/main/java/com/sflpro/identity/core/services/principal/PrincipalServiceImpl.java [hunk 19]
```diff
@@ -25,27 +27,57 @@
...
+    private Principal upsert(final Identity identity, final PrincipalUpdateRequest updateRequest) {
         Assert.notNull(identity, "identity cannot be null");
-        Assert.notNull(identity.getId(), "identity id cannot be null");
-        List<Principal> principals = principalRepository.findAllByDeletedIsNullAndIdentity(identity);
-        if (principals.isEmpty()) {
-            throw new ResourceNotFoundException(String.format("Principal with identity:%s not found", identity.getId()));
+        Assert.notNull(updateRequest, "updateRequest cannot be null");
+        try {
+            Principal principal = get(updateRequest.getPrincipalType(), updateRequest.getPrincipalName());
+            if (!principal.getIdentity().equals(identity)) {
```
Comment (dnavre) [inline identity-core/identity-core-services-impl/src/main/java/com/sflpro/identity/core/services/principal/PrincipalServiceImpl.java]: Verify that the equals() method is actually something that satisfies your needs here.

File: identity-core/identity-core-services-impl/src/main/java/com/sflpro/identity/core/services/principal/PrincipalServiceImpl.java [hunk 20]
```diff
@@ -25,27 +27,57 @@
...
+                throw new PrincipalNameBusyException(updateRequest.getPrincipalType(), updateRequest.getPrincipalName());
+            }
+            principal.setName(updateRequest.getPrincipalName());
+            principal.setPrincipalType(updateRequest.getPrincipalType());
+            principal.setPrincipalStatus(updateRequest.getPrincipalStatus());
+            return principalRepository.save(principal);
+        } catch (ResourceNotFoundException e) {
```
Comment (dnavre) [inline identity-core/identity-core-services-impl/src/main/java/com/sflpro/identity/core/services/principal/PrincipalServiceImpl.java]: This catch close should be written exactly around the line that should produce the exception.

File: identity-core/identity-core-services/src/main/java/com/sflpro/identity/core/services/identity/IdentityService.java [hunk 21]
```diff
@@ -26,20 +29,33 @@
      * @param identityCreationRequest identity creation request model
      * @return id and details of the created identity
      */
+    @Deprecated
+    @Transactional
```
Comment (dnavre) [inline identity-core/identity-core-services/src/main/java/com/sflpro/identity/core/services/identity/IdentityService.java]: Remove @Transactional annotations from here.

File: identity-core/identity-core-services-impl/src/main/java/com/sflpro/identity/core/services/identity/IdentityServiceImpl.java [hunk 22]
```diff
@@ -157,6 +189,18 @@ protected Identity changeSecret(final Identity identity, final String newSecret)
         return identityRepository.save(identity);
     }
 
+    @Override
+    public boolean chkSecretCorrectAndIdentityActive(final Identity identity, final String secret) throws AuthenticationServiceException {
+        if (!secretHashHelper.isSecretCorrect(secret, identity.getSecret())) {
+            logger.debug("Invalid secret was supplied for identity {}.", identity);
+            throw new AuthenticationServiceException("Invalid credentials");
+        } else if (!(identity.getStatus() == IdentityStatus.ACTIVE)) {
```
Comment (sfl-sonar) [inline identity-core/identity-core-services-impl/src/main/java/com/sflpro/identity/core/services/identity/IdentityServiceImpl.java]: MINOR Use the opposite operator ("!=") instead. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1940)

File: identity-core/identity-core-services/src/main/java/com/sflpro/identity/core/services/identity/IdentityService.java [hunk 23]
```diff
@@ -26,21 +27,39 @@
      * @param identityCreationRequest identity creation request model
      * @return id and details of the created identity
      */
+    @Deprecated
     Identity create(final IdentityCreationRequest identityCreationRequest);
```
Comment (sfl-sonar) [inline identity-core/identity-core-services/src/main/java/com/sflpro/identity/core/services/identity/IdentityService.java]: MAJOR Add the missing @deprecated Javadoc tag. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AMissingDeprecatedCheck)
INFO Do not forget to remove this deprecated code someday. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1133)

**Comment by sfl-sonar** [comment]
SonarQube analysis reported 11 issues
* ![MAJOR][MAJOR] 6 major
* ![MINOR][MINOR] 1 minor
* ![INFO][INFO] 4 info

Watch the comments in this conversation to review them.

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by sfl-sonar** [comment]
SonarQube analysis reported 10 issues
* ![MAJOR][MAJOR] 6 major
* ![INFO][INFO] 4 info

Watch the comments in this conversation to review them.

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: identity-core/identity-core-services-impl/src/main/java/com/sflpro/identity/core/services/principal/PrincipalServiceImpl.java [hunk 24]
```diff
@@ -21,31 +31,85 @@
...
+     */
     @Override
-    public List<Principal> getByIdentity(final Identity identity) {
+    @Transactional
+    public void deleteAllByIdentity(final Identity identity) {
+        Assert.notNull(identity, "identity cannot be null");
+        logger.debug("Deleting principals by identity id {}", identity.getId());
+        Iterable<Principal> principals = principalRepository.findAllByIdentity(identity);
+        LocalDateTime now = LocalDateTime.now();
+        principals.forEach(p -> p.setDeleted(now)); // TODO chk with Mr. Smith do we need also invalidate tokens issued by this principals
```
Comment (sfl-sonar) [inline identity-core/identity-core-services-impl/src/main/java/com/sflpro/identity/core/services/principal/PrincipalServiceImpl.java]: INFO Complete the task associated to this TODO comment. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1135)

**Comment by sfl-sonar** [comment]
SonarQube analysis reported 11 issues
* ![MAJOR][MAJOR] 6 major
* ![INFO][INFO] 5 info

Watch the comments in this conversation to review them.

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: identity-api/identity-api-facade/src/main/java/com/sflpro/identity/api/endpoints/IdentityEndpoint.java [hunk 25]
```diff
@@ -74,21 +82,35 @@ public IdentityCreationResponseDto create(@NotNull @Valid final IdentityCreation
...
     }
 
     @ApiOperation("Delete identity")
     @DELETE
     @Path("/{identityId}")
-    public IdentityUpdateResponseDto delete(@NotNull @PathParam("identityId") final String identityId) {
-        return new IdentityUpdateResponseDto();
+    @Deprecated
+    @Transactional
+    public IdentityDto delete(@NotNull @PathParam("identityId") final String identityId) {
```
Comment (sfl-sonar) [inline identity-api/identity-api-facade/src/main/java/com/sflpro/identity/api/endpoints/IdentityEndpoint.java]: MAJOR Add the missing @deprecated Javadoc tag. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AMissingDeprecatedCheck)
INFO Do not forget to remove this deprecated code someday. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1133)

File: identity-core/identity-core-services-impl/src/main/java/com/sflpro/identity/core/services/principal/PrincipalServiceImpl.java [hunk 26]
```diff
@@ -21,31 +31,85 @@
...
+     */
     @Override
-    public List<Principal> getByIdentity(final Identity identity) {
+    @Transactional
+    public void deleteAllByIdentity(final Identity identity) {
+        Assert.notNull(identity, "identity cannot be null");
+        logger.debug("Deleting principals by identity id {}", identity.getId());
+        Iterable<Principal> principals = principalRepository.findAllByIdentity(identity);
+        LocalDateTime now = LocalDateTime.now();
+        principals.forEach(p -> p.setDeleted(now));
```
Comment (sfl-sonar) [inline identity-core/identity-core-services-impl/src/main/java/com/sflpro/identity/core/services/principal/PrincipalServiceImpl.java]: INFO Complete the task associated to this TODO comment. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1135)

**Comment by sfl-sonar** [comment]
SonarQube analysis reported 10 issues
* ![MAJOR][MAJOR] 6 major
* ![INFO][INFO] 4 info

Watch the comments in this conversation to review them.

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: identity-core/identity-core-services-impl/src/main/java/com/sflpro/identity/core/services/identity/IdentityServiceImpl.java [hunk 27]
```diff
@@ -68,7 +71,12 @@ public Identity get(final String identityId) {
         return identity;
     }
 
+    /**
+     * {@inheritDoc}
+     */
     @Override
+    @Deprecated
+    @Transactional
     public Identity create(final IdentityCreationRequest identityCreationRequest) {
```
Comment (sfl-sonar) [inline identity-core/identity-core-services-impl/src/main/java/com/sflpro/identity/core/services/identity/IdentityServiceImpl.java]: MAJOR Add the missing @deprecated Javadoc tag. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AMissingDeprecatedCheck)
INFO Do not forget to remove this deprecated code someday. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1133)

**Comment by sfl-sonar** [comment]
SonarQube analysis reported 11 issues
* ![MAJOR][MAJOR] 7 major
* ![INFO][INFO] 4 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [IdentityServiceImpl.java#L189](https://github.com/sflpro/identity/blob/f9c03553cbd8fd55b153c338cb71724afb002e77/identity-core/identity-core-services-impl/src/main/java/com/sflpro/identity/core/services/identity/IdentityServiceImpl.java#L189): Make this method "public" or remove the "@Transactional" annotation [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2230)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Pull request merged**